File: backend/faos_core/agents/base_agent.py
"""FAOS v6.0 — Base agent executor with JSON fallback (Python)."""

from __future__ import annotations

import json
import time
from typing import Any, Dict, Optional

from faos_core.orchestrator.agent_registry import get_agent, set_agent_state


def _safe_parse_json(raw: str) -> Optional[Dict[str, Any]]:
    try:
        parsed = json.loads(raw)
        if isinstance(parsed, dict):
            return parsed
        return {"value": parsed}
    except (json.JSONDecodeError, TypeError):
        return None


def build_deterministic_deliverable(
    agent: Dict[str, Any], input_data: Dict[str, Any]
) -> Dict[str, Any]:
    command = str(input_data.get("command") or "")
    return {
        "agent_id": agent.get("id"),
        "agent_name": agent.get("name"),
        "domain": agent.get("domain"),
        "task_id": input_data.get("task_id"),
        "category": input_data.get("category") or "general",
        "brand": input_data.get("brand") or "BulletsEye / FMK Group",
        "summary": f"{agent.get('name')} processed: {command[:180]}",
        "deliverable": "\n".join(
            [
                f"# {agent.get('name')} Output",
                "",
                f"Domain: {agent.get('domain')}",
                f"Brief: {command}",
                "",
                "## Recommended actions",
                "1. Confirm scope with client",
                "2. Produce draft deliverable for QA",
                "3. Route to approval gate",
            ]
        ),
        "model_hint": agent.get("model_hint"),
        "status": "completed",
    }


def run_agent_module(
    agent_id: str,
    input_data: Dict[str, Any],
    *,
    enrich_raw: Optional[str] = None,
) -> Dict[str, Any]:
    started = time.time()
    agent = get_agent(agent_id)

    if not agent:
        state = set_agent_state(
            agent_id,
            status="failed",
            current_task=input_data.get("task_id"),
            output=None,
            error="agent_not_registered",
            latency_ms=(time.time() - started) * 1000,
        )
        return {
            "state": state,
            "deliverable": f"Agent {agent_id} is not registered.",
            "structured": {"error": "agent_not_registered"},
            "used_fallback": True,
        }

    set_agent_state(
        agent_id,
        status="running",
        current_task=input_data.get("task_id"),
        output=None,
        error=None,
    )

    used_fallback = False
    structured = build_deterministic_deliverable(agent, input_data)

    if enrich_raw is not None:
        parsed = _safe_parse_json(enrich_raw)
        if parsed is None:
            used_fallback = True
            structured = {
                **structured,
                "llm_raw": enrich_raw[:4000],
                "note": "malformed_json_fallback",
            }
        else:
            structured = {**structured, **parsed, "llm_enriched": True}

    deliverable = (
        structured["deliverable"]
        if isinstance(structured.get("deliverable"), str)
        else json.dumps(structured, indent=2)
    )

    state = set_agent_state(
        agent_id,
        status="fallback" if used_fallback else "completed",
        current_task=input_data.get("task_id"),
        output=structured,
        error=str(structured.get("note")) if used_fallback else None,
        latency_ms=(time.time() - started) * 1000,
    )

    return {
        "state": state,
        "deliverable": deliverable,
        "structured": structured,
        "used_fallback": used_fallback,
    }

File: faos_core/agents/base-agent.ts
/**
 * FAOS v6.0 — Base agent executor with JSON fallback
 */

import {
  getAgent,
  setAgentState,
  type RegisteredAgent,
} from "@/faos_core/orchestrator/agent-registry";
import type { AgentState } from "@/faos_core/types";

export type AgentTaskInput = {
  task_id: string;
  command: string;
  category?: string;
  brand?: string;
  metadata?: Record<string, unknown>;
};

export type AgentRunResult = {
  state: AgentState;
  deliverable: string;
  structured: Record<string, unknown>;
  used_fallback: boolean;
};

function safeParseJson(raw: string): Record<string, unknown> | null {
  try {
    const parsed: unknown = JSON.parse(raw);
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
      return parsed as Record<string, unknown>;
    }
    return { value: parsed as unknown };
  } catch {
    return null;
  }
}

export function buildDeterministicDeliverable(
  agent: RegisteredAgent,
  input: AgentTaskInput
): Record<string, unknown> {
  return {
    agent_id: agent.id,
    agent_name: agent.name,
    domain: agent.domain,
    task_id: input.task_id,
    category: input.category || "general",
    brand: input.brand || "BulletsEye / FMK Group",
    summary: `${agent.name} processed: ${input.command.slice(0, 180)}`,
    deliverable: [
      `# ${agent.name} Output`,
      ``,
      `Domain: ${agent.domain}`,
      `Brief: ${input.command}`,
      ``,
      `## Recommended actions`,
      `1. Confirm scope with client`,
      `2. Produce draft deliverable for QA`,
      `3. Route to approval gate`,
    ].join("\n"),
    model_hint: agent.model_hint,
    status: "completed",
  };
}

/**
 * Run a single agent with timeout + malformed-JSON fallback.
 * LLM enrichment is optional; deterministic path always succeeds.
 */
export async function runAgentModule(
  agentId: string,
  input: AgentTaskInput,
  options?: {
    timeout_ms?: number;
    enrich?: (agent: RegisteredAgent, input: AgentTaskInput) => Promise<string>;
  }
): Promise<AgentRunResult> {
  const agent = getAgent(agentId);
  const started = Date.now();

  if (!agent) {
    const state = setAgentState(agentId, {
      status: "failed",
      current_task: input.task_id,
      output: null,
      error: "agent_not_registered",
      latency_ms: Date.now() - started,
    });
    return {
      state,
      deliverable: `Agent ${agentId} is not registered.`,
      structured: { error: "agent_not_registered" },
      used_fallback: true,
    };
  }

  setAgentState(agentId, {
    status: "running",
    current_task: input.task_id,
    output: null,
    error: undefined,
  });

  const timeoutMs = options?.timeout_ms ?? 12_000;
  let usedFallback = false;
  let structured: Record<string, unknown>;

  try {
    const work = (async () => {
      if (options?.enrich) {
        const raw = await options.enrich(agent, input);
        const parsed = safeParseJson(raw);
        if (!parsed) {
          usedFallback = true;
          const base = buildDeterministicDeliverable(agent, input);
          return {
            ...base,
            llm_raw: raw.slice(0, 4000),
            note: "malformed_json_fallback",
          };
        }
        return {
          ...buildDeterministicDeliverable(agent, input),
          ...parsed,
          llm_enriched: true,
        };
      }
      return buildDeterministicDeliverable(agent, input);
    })();

    structured = await Promise.race([
      work,
      new Promise<Record<string, unknown>>((_, reject) => {
        setTimeout(() => reject(new Error("agent_timeout")), timeoutMs);
      }),
    ]);
  } catch (err) {
    usedFallback = true;
    const message = err instanceof Error ? err.message : "agent_failed";
    structured = {
      ...buildDeterministicDeliverable(agent, input),
      note: message === "agent_timeout" ? "timeout_fallback" : "error_fallback",
      error: message,
    };
  }

  const deliverable =
    typeof structured.deliverable === "string"
      ? structured.deliverable
      : JSON.stringify(structured, null, 2);

  const state = setAgentState(agentId, {
    status: usedFallback ? "fallback" : "completed",
    current_task: input.task_id,
    output: structured,
    error: usedFallback ? String(structured.note || "fallback") : undefined,
    latency_ms: Date.now() - started,
  });

  return { state, deliverable, structured, used_fallback: usedFallback };
}
